src/app/clusters/icd-management-server/icd-management-server.cpp

```diff
@@ -67,43 +72,173 @@ CHIP_ERROR IcdManagementAttributeAccess::Read(const ConcreteReadAttributePath &
+        auto storage = &chip::Server::GetInstance().GetPersistentStorage();
+
+        IcdMonitoringTable table(storage, fabric);
+        IcdMonitoringEntry e;
+        uint16_t i     = 0;
+        CHIP_ERROR err = CHIP_NO_ERROR;
+
+        while ((CHIP_NO_ERROR == err) && (i < table.Limit()))
+        {
+            err = table.Get(i++, e);
+            if (CHIP_NO_ERROR == err)
+            {
+                ReturnErrorOnFailure(subEncoder.Encode(e));
+            }
+        }
+        return (CHIP_ERROR_NOT_FOUND == err) ? CHIP_NO_ERROR : err;
+    });
+}
+
+InteractionModel::Status CheckAdmin(chip::app::CommandHandler * commandObj, bool & is_admin)
+{
+    Access::SubjectDescriptor desc = commandObj->GetSubjectDescriptor();
+    chip::Access::AccessControl::EntryIterator iterator;
+    chip::Access::AccessControl::Entry entry;
+    chip::Access::Privilege priv;
+
+    CHIP_ERROR err = GetAccessControl().Entries(commandObj->GetAccessingFabricIndex(), iterator);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, InteractionModel::Status::Failure);
+
+    is_admin = false;
+    while ((err = iterator.Next(entry)) == CHIP_NO_ERROR)
+    {
+        size_t count = 0;
+        entry.GetPrivilege(priv);
+        entry.GetSubjectCount(count);
+        for (size_t i = 0; !is_admin && Privilege::kAdminister == priv && i < count; ++i)
+        {
+            NodeId id;
+            entry.GetSubject(i, id);
+            is_admin = (desc.subject == id);
+        }
+    }
+
+    return InteractionModel::Status::Success;
 }
 
 IcdManagementAttributeAccess gAttribute;
 
 } // namespace
 
-//==============================================================================
-
 InteractionModel::Status
 IcdManagementServer::RegisterClient(chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
                                     const chip::app::Clusters::IcdManagement::Commands::RegisterClient::DecodableType & commandData)
 {
-    // TODO: Implementent logic for end device
-    return InteractionModel::Status::UnsupportedCommand;
+    FabricIndex fabric = commandObj->GetAccessingFabricIndex();
+    IcdMonitoringTable table(&chip::Server::GetInstance().GetPersistentStorage(), fabric);
+
+    // Get current entry, if exists
+    IcdMonitoringEntry entry;
+    CHIP_ERROR err = table.Find(commandData.checkInNodeID, entry);
+    if (CHIP_NO_ERROR == err)
+    {
+        // Existing entry, check admin status
+        bool is_admin                   = false;
+        InteractionModel::Status status = CheckAdmin(commandObj, is_admin);
+        VerifyOrReturnError(InteractionModel::Status::Success == status, status);
+
+        if (is_admin)
+        {
+            // Administrator
+            if (commandData.verificationKey.HasValue())
+            {
+                VerifyOrReturnError(commandData.verificationKey.Value().data_equal(entry.key), InteractionModel::Status::Failure);
+            }
```

Comment:
Spec text for admin doesn't check the VerificationKey even if one is provided by the Admin. Should we update the spec or the implementation?

Reply:
The verification key is now ignored if the ISD has administrator permissions.